File: copy_coze_token.py
#!/usr/bin/env python3
from __future__ import annotations
import base64
import hashlib
import hmac
import json
import os
import re
import socket
from pathlib import Path
from typing import Any, Dict, Tuple

ROOT = Path(__file__).resolve().parent
TOKEN_FILE = ROOT / 'config' / '.copy-coze-token.secure'
SESSION_TOKEN = ''
SECURE_VERSION = 'local_encrypted_v1'


def normalize_token(value: Any) -> str:
    token = str(value or '').replace('\ufeff', '')
    token = re.sub(r'[\u200B-\u200D\u2060\uFEFF]', '', token).strip()
    token = re.sub(r'^Bearer\s+', '', token, flags=re.I).strip()
    if len(token) >= 2 and token[0] == token[-1] and token[0] in ('"', "'"):
        token = token[1:-1].strip()
    return re.sub(r'\s+', '', token)


def _machine_secret() -> bytes:
    user = os.environ.get('USERNAME') or os.environ.get('USER') or 'user'
    machine = os.environ.get('COMPUTERNAME') or socket.gethostname() or 'machine'
    return f'{user}|{machine}|copy-coze-token-v1'.encode('utf-8')


def _derive_key(salt: bytes) -> bytes:
    return hashlib.pbkdf2_hmac('sha256', _machine_secret(), salt, 120000, dklen=32)


def _xor_stream(data: bytes, key: bytes, iv: bytes) -> bytes:
    out = bytearray(len(data)); offset = 0; counter = 0
    while offset < len(data):
        block = hmac.new(key, iv + counter.to_bytes(4, 'big'), hashlib.sha256).digest(); counter += 1
        for value in block:
            if offset >= len(data): break
            out[offset] = data[offset] ^ value; offset += 1
    return bytes(out)


def _encrypt_token(token: str) -> str:
    salt, iv = os.urandom(16), os.urandom(16)
    key = _derive_key(salt)
    cipher = _xor_stream(token.encode('utf-8'), key, iv)
    tag = hmac.new(key, iv + cipher, hashlib.sha256).digest()
    return json.dumps({'version': 1, 'scheme': 'PBKDF2-HMAC-SHA256-STREAM', 'salt': base64.b64encode(salt).decode(), 'iv': base64.b64encode(iv).decode(), 'tag': base64.b64encode(tag).decode(), 'data': base64.b64encode(cipher).decode()}, ensure_ascii=False)


def _decrypt_token(text: str) -> str:
    obj = json.loads(text)
    if not isinstance(obj, dict) or obj.get('version') != 1:
        raise RuntimeError('令牌文件版本不受支持')
    salt = base64.b64decode(obj.get('salt') or ''); iv = base64.b64decode(obj.get('iv') or '')
    tag = base64.b64decode(obj.get('tag') or ''); cipher = base64.b64decode(obj.get('data') or '')
    key = _derive_key(salt); actual = hmac.new(key, iv + cipher, hashlib.sha256).digest()
    if not hmac.compare_digest(tag, actual):
        raise RuntimeError('令牌文件校验失败，可能来自其他 Windows 用户或电脑')
    return normalize_token(_xor_stream(cipher, key, iv).decode('utf-8'))


def _stored_token() -> str:
    if not TOKEN_FILE.exists(): return ''
    try: return _decrypt_token(TOKEN_FILE.read_text('utf-8'))
    except Exception: return ''


def get_token(private_config: Dict[str, Any]) -> Tuple[str, str]:
    global SESSION_TOKEN
    current = normalize_token(SESSION_TOKEN)
    if current: return current, 'session'
    stored = _stored_token()
    if stored: return stored, SECURE_VERSION
    env = normalize_token(os.environ.get('COZE_API_TOKEN') or '')
    if env: return env, 'environment'
    legacy = normalize_token((private_config or {}).get('token') or '')
    if legacy: return legacy, 'legacy_config'
    return '', 'not_configured'


def _source_label(source: str) -> str:
    return {'session': '仅本次运行', 'environment': '环境变量', SECURE_VERSION: '本机加密保存', 'legacy_config': '旧版私有配置', 'not_configured': '未配置'}.get(source, '未配置')


def status_payload(private_config: Dict[str, Any]) -> Dict[str, Any]:
    token, source = get_token(private_config)
    return {'ok': True, 'configured': bool(token), 'tokenLoaded': bool(token), 'tokenSource': source, 'tokenSourceLabel': _source_label(source), 'secureStorage': SECURE_VERSION, 'manualTokenInput': True, 'message': ('扣子访问令牌已配置 · ' + _source_label(source)) if token else '请手动输入有效的扣子访问令牌'}


def configure_token(payload: Dict[str, Any], private_config: Dict[str, Any]) -> Dict[str, Any]:
    global SESSION_TOKEN
    payload = payload if isinstance(payload, dict) else {}
    if payload.get('clear'):
        SESSION_TOKEN = ''
        try: TOKEN_FILE.unlink(missing_ok=True)
        except Exception: pass
        out = status_payload(private_config); out['cleared'] = True; return out
    raw = str(payload.get('token') or '')
    token = normalize_token(raw)
    if not token: return {'ok': False, 'error': {'code': 'token_missing', 'message': '请输入扣子访问令牌'}}
    if re.match(r'^https?://', token, re.I) or len(token) < 20: return {'ok': False, 'error': {'code': 'token_invalid', 'message': '令牌格式或长度异常，请从扣子平台重新复制完整令牌'}}
    if re.match(r'^(?:your[_-]?token|replace[_-]?me|token)$', token, re.I): return {'ok': False, 'error': {'code': 'token_placeholder', 'message': '当前内容是占位文本，不是有效令牌'}}
    SESSION_TOKEN = token
    persisted = False; warning = ''
    if payload.get('remember'):
        try:
            TOKEN_FILE.parent.mkdir(parents=True, exist_ok=True)
            TOKEN_FILE.write_text(_encrypt_token(token), 'utf-8')
            persisted = True
            SESSION_TOKEN = ''
        except Exception as exc:
            warning = '令牌已在本次运行中生效，但本机加密保存失败：' + str(exc)
    out = status_payload(private_config)
    out.update({'saved': True, 'persisted': persisted, 'normalized': token != raw.strip(), 'warning': warning})
    return out

File: copy-coze-token.js
'use strict';
const fs=require('fs');
const path=require('path');
const crypto=require('crypto');

const SECURE_VERSION='local_encrypted_v1';

module.exports=function createCopyCozeTokenStore({root,privateConfig}){
  const secureFile=path.join(root,'config','.copy-coze-token.secure');
  let sessionToken='';

  function normalizeToken(value){
    let token=String(value==null?'':value)
      .replace(/^\uFEFF/,'')
      .replace(/[\u200B-\u200D\u2060\uFEFF]/g,'')
      .trim();
    token=token.replace(/^Bearer\s+/i,'').trim();
    if((token.startsWith('"')&&token.endsWith('"'))||(token.startsWith("'")&&token.endsWith("'")))token=token.slice(1,-1).trim();
    return token.replace(/\s+/g,'');
  }
  function machineSecret(){
    return [process.env.USERNAME||process.env.USER||'user',process.env.COMPUTERNAME||process.env.HOSTNAME||'machine','copy-coze-token-v1'].join('|');
  }
  function deriveKey(salt){return crypto.pbkdf2Sync(machineSecret(),salt,120000,32,'sha256');}
  function xorStream(data,key,iv){
    const out=Buffer.alloc(data.length);let offset=0,counter=0;
    while(offset<data.length){
      const c=Buffer.alloc(4);c.writeUInt32BE(counter++);
      const block=crypto.createHmac('sha256',key).update(iv).update(c).digest();
      for(let i=0;i<block.length&&offset<data.length;i++,offset++)out[offset]=data[offset]^block[i];
    }
    return out;
  }
  function encryptToken(token){
    const salt=crypto.randomBytes(16),iv=crypto.randomBytes(16),key=deriveKey(salt),plain=Buffer.from(token,'utf8');
    const cipher=xorStream(plain,key,iv),tag=crypto.createHmac('sha256',key).update(iv).update(cipher).digest();
    return JSON.stringify({version:1,scheme:'PBKDF2-HMAC-SHA256-STREAM',salt:salt.toString('base64'),iv:iv.toString('base64'),tag:tag.toString('base64'),data:cipher.toString('base64')});
  }
  function decryptToken(text){
    const obj=JSON.parse(String(text||''));
    if(!obj||obj.version!==1)throw new Error('令牌文件版本不受支持');
    const salt=Buffer.from(obj.salt||'','base64'),iv=Buffer.from(obj.iv||'','base64'),tag=Buffer.from(obj.tag||'','base64'),cipher=Buffer.from(obj.data||'','base64'),key=deriveKey(salt);
    const actual=crypto.createHmac('sha256',key).update(iv).update(cipher).digest();
    if(tag.length!==actual.length||!crypto.timingSafeEqual(tag,actual))throw new Error('令牌文件校验失败，可能来自其他 Windows 用户或电脑');
    return normalizeToken(xorStream(cipher,key,iv).toString('utf8'));
  }
  function storedToken(){
    if(!fs.existsSync(secureFile))return'';
    try{return decryptToken(fs.readFileSync(secureFile,'utf8'));}catch(_e){return'';}
  }
  function getToken(){
    const current=normalizeToken(sessionToken);if(current)return{token:current,source:'session'};
    const stored=storedToken();if(stored)return{token:stored,source:SECURE_VERSION};
    const env=normalizeToken(process.env.COZE_API_TOKEN||'');if(env)return{token:env,source:'environment'};
    const legacy=normalizeToken((privateConfig||{}).token||'');if(legacy)return{token:legacy,source:'legacy_config'};
    return{token:'',source:'not_configured'};
  }
  function sourceLabel(source){
    return {session:'仅本次运行',environment:'环境变量',local_encrypted_v1:'本机加密保存',legacy_config:'旧版私有配置',not_configured:'未配置'}[source]||'未配置';
  }
  function status(){
    const t=getToken();
    return{ok:true,configured:!!t.token,tokenLoaded:!!t.token,tokenSource:t.source,tokenSourceLabel:sourceLabel(t.source),secureStorage:SECURE_VERSION,manualTokenInput:true,message:t.token?('扣子访问令牌已配置 · '+sourceLabel(t.source)):'请手动输入有效的扣子访问令牌'};
  }
  function configure(payload){
    payload=payload&&typeof payload==='object'?payload:{};
    if(payload.clear){
      sessionToken='';
      try{fs.unlinkSync(secureFile);}catch(_e){}
      return Object.assign(status(),{cleared:true});
    }
    const raw=String(payload.token==null?'':payload.token),token=normalizeToken(raw);
    if(!token)return{ok:false,error:{code:'token_missing',message:'请输入扣子访问令牌'}};
    if(/^https?:\/\//i.test(token)||token.length<20)return{ok:false,error:{code:'token_invalid',message:'令牌格式或长度异常，请从扣子平台重新复制完整令牌'}};
    if(/^(?:your[_-]?token|replace[_-]?me|token)$/i.test(token))return{ok:false,error:{code:'token_placeholder',message:'当前内容是占位文本，不是有效令牌'}};
    sessionToken=token;
    let persisted=false,warning='';
    if(payload.remember){
      try{
        fs.mkdirSync(path.dirname(secureFile),{recursive:true});
        fs.writeFileSync(secureFile,encryptToken(token),{encoding:'utf8',mode:0o600});
        persisted=true;
        sessionToken='';
      }catch(e){warning='令牌已在本次运行中生效，但本机加密保存失败：'+e.message;}
    }
    return Object.assign(status(),{saved:true,persisted,normalized:token!==raw.trim(),warning});
  }
  return{getToken,status,configure,normalizeToken,secureFile,SECURE_VERSION};
};
